### PR TITLE
Jinghang's Binary Converter

### PR DESCRIPTION
**Pull request by Python123434**
Code to convert a binary string into an integer.

### CODE REVIEW

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package JinghangCodeLab;
+
+public class MagicNumberConverter {
+
+    public static int binaryToDecimal(String binary)
+    {  
+        //edge cases
+        if (binary == "0"){
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java]: coding style: add a space between ")" and "{".

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package JinghangCodeLab;
+
+public class MagicNumberConverter {
+
+    public static int binaryToDecimal(String binary)
+    {  
+        //edge cases
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java]: coding style: for comments, typically we leave a space after "//"

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            char place = binary.charAt(i);
+            int num = Character.getNumericValue(place);
+            int val = multiple * num;
+            sum += val;
+            multiple *= 2;
+        }
+        if(isNegative == 0){
+            return sum*-1;
+        }
+
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java:47]: nit: remove the empty lines, these lines make it harder to read the code.

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        //negative
+        int isNegative = -1;
+        if(Integer.parseInt(binary) < 0){
+            isNegative = 0;
+        }
+        else{
+            isNegative = -1;
+        }
+        //formula
+        int length =  binary.length() - 1;
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java]: coding style: only keep one space between "=" and "binary.length()".

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    public static int binaryToDecimal(String binary)
+    {  
+        //edge cases
+        if (binary == "0"){
+            return 0;
+        }
+
+        //negative
+        int isNegative = -1;
+        if(Integer.parseInt(binary) < 0){
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java]: It's great that you thought of handling negative numbers, however, using "Integer.parseInt" may cause overflow issue, for example if the binary has 20 characters like this:  "11111111111111111111". You can try to check if the first character of the input binary is "-".

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    public static int binaryToDecimal(String binary)
+    {  
+        //edge cases
+        if (binary == "0"){
+            return 0;
+        }
+
+        //negative
+        int isNegative = -1;
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java]: isNegative variable can be a boolean type instead of "int" type. making it "boolean" type will improve code's readability.

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        if(Integer.parseInt(binary) < 0){
+            isNegative = 0;
+        }
+        else{
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java]: In java, for if/else statement, please following the below coding convention: 

```
if (condition1) {
    // code to execute if condition1 is true
} else if (condition2) {
    // code to execute if condition1 is false and condition2 is true
} else if (condition3) {
    // code to execute if condition1 and condition2 are false and condition3 is true
} else {
    // code to execute if none of the conditions are true
}

```

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+            isNegative = -1;
+        }
+        //formula
+        int length =  binary.length() - 1;
+        int multiple = 1;
+        int sum = 0;
+        for (int i = length; i > isNegative; i--){
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java]: i > isNegative is a little bit confusing and not easy to understand. Try to create another variable to keep track of the index of the digits.

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            char place = binary.charAt(i);
+            int num = Character.getNumericValue(place);
+            int val = multiple * num;
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java]: it will improve cod readability to combine them into one line: 

int val = multiple * Character.getNumericValue(binary.chatAt(i));

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+            sum += val;
+            multiple *= 2;
+        }
+        if(isNegative == 0){
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java]: coding style: add a space between "if" and "(", same for ")" and "{"

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+            return sum*-1;
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java]: coding style: add space, e.g. "return sum * -1;", or simply "return -sum;"

File: codelab/JinghangCodeLab/MagicNumberConverterTest.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+public class MagicNumberConverterTest {
+    @Test
+    public void testConvertBinaryNumber() {
+        MagicNumberConverter convert = new MagicNumberConverter();
+                assertEquals(-10,convert.binaryToDecimal("-1010"));
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverterTest.java]: add more unit tests cases to test other possible inputs, for example: "11111", "0", "1".

long string input such as "11111111111111111111".

consider other corner cases such as empty string as input, or when the string is very long, for example, 50 digits, how do you handle such cases?

Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverterTest.java]: coding style: indent is incorrect.

**Review by paulchen503 — CHANGES_REQUESTED** [review]
Overall looks good, please address coding styles issue and add more unit tests.

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
+package JinghangCodeLab;
+
+public class MagicNumberConverter {
+
+    public static int binaryToDecimal(String binary) {
+        if (Double.parseDouble(binary) > Integer.MAX_VALUE){
+            System.out.println("Out of Range not in Integer Range");
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java]: you can thrown an exception in this case.

throw new OutOfRangeException("Input binary number is of out of range, input binary number=" + binary);

File: codelab/JinghangCodeLab/MagicNumberConverter.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+            isNegative = true;
+        } else {
+            isNegative = false;
+        }
+
+        // Formula
+        int length = binary.length() - 1;
+        int multiple = 1;
+        int sum = 0;
+        int subtractIfNegative;
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverter.java:33]: you can also write in this way, then you don't need the if/else statement.

`int subtractIfNegative = (isNegative ? 0 : -1);`

File: codelab/JinghangCodeLab/MagicNumberConverterTest.java [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+public class MagicNumberConverterTest {
+    @Test
+    public void testConvertBinaryNumber() {
+        MagicNumberConverter convert = new MagicNumberConverter();
+        assertEquals(0, convert.binaryToDecimal("11111111111111111111111111111111111111111111111111"));
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverterTest.java:11]: create a separate test method for this sceanrio, "testNumberOutOfRange". you can use the following to verify that an exception is thrown for too large number:

`assertThrows(OutOfRangeException.class, convert.binaryToDecimal("11111111111111111111111111111111111111111111111"))`

File: codelab/JinghangCodeLab/MagicNumberConverterTest.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
+package JinghangCodeLab;
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+public class MagicNumberConverterTest {
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverterTest.java:7]: add a test case for empty string input, what should be the expected behavior for empty string? I recommend throw an IllegalArgumentException for this case. 

convert.binaryToDecimal("")

File: codelab/JinghangCodeLab/MagicNumberConverterTest.java [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+    public void testNegative() {
+        MagicNumberConverter convert = new MagicNumberConverter();
+        assertEquals(-10, convert.binaryToDecimal("-1010"));
+    }
+    
+    @Test
+    public void testDecimal() {
+        MagicNumberConverter convert = new MagicNumberConverter();
+        assertEquals(0, convert.binaryToDecimal("0.1010001"));
+
```
Comment (paulchen503) [inline codelab/JinghangCodeLab/MagicNumberConverterTest.java:26]: remove empty line.

**Review by paulchen503 — COMMENTED** [review]
Second iteration looks much better, great work, left some additional comments for more improvements.

**Review by paulchen503 — APPROVED** [review]
Looks good to me, nice work!
